### PR TITLE
A.18: Assignments

### PR DESCRIPTION
**Pull request by asfar-nazar**

### CODE REVIEW

File: Parser/Token.cs [hunk 1]
```diff
@@ -37,9 +37,24 @@ public enum E {
 
    // Utility function used to echo an error to the console
    public void PrintError () {
+      Console.OutputEncoding = Encoding.Unicode;
       if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
       Console.ForegroundColor = ConsoleColor.Yellow;
```
Comment (tarydon) [inline Parser/Token.cs:56]: See the sample image I sent you - only the actual error text is in yellow, the source code is printed in white.

File: Parser/Token.cs [hunk 2]
```diff
@@ -37,9 +37,24 @@ public enum E {
...
    public void PrintError () {
+      Console.OutputEncoding = Encoding.Unicode;
       if (Kind != ERROR) throw new Exception ("PrintError called on a non-error token");
       Console.ForegroundColor = ConsoleColor.Yellow;
-      Console.WriteLine ($"At line {Line}, column {Column} of {Source.FileName}: {Text}");
+      Console.Write ($"File: {Source.FileName}");
+      var cPos = Console.CursorLeft;
+      Console.WriteLine ($"\n\u2500\u2500\u2500\u252c{string.Join ("", Enumerable.Repeat ("\u2500", cPos - 3))}");
+      var lines = Source.Lines;
+      if (Line < lines.Length - 3) Console.WriteLine ($"{Line - 2,3}\u2502{lines[Line - 3]}");
```
Comment (tarydon) [inline Parser/Token.cs]: This repeated logic (Line < lines.Length - _n_) is a bit clumsy. Try to generalize it with a loop of some kind. In general, we want to be able to easily customize it to include a different number of lines of context. A helper routine like `int Clamp (this int n, int min, int max)` could be useful.

File: Parser/Token.cs [hunk 3]
```diff
@@ -37,9 +37,24 @@ public enum E {
...
+      if (Line < lines.Length - 2) Console.WriteLine ($"{Line - 1,3}\u2502{lines[Line - 2]}");
+      if (Line < lines.Length - 1) Console.WriteLine ($"{Line,3}\u2502{lines[Line - 1]}");
+      cPos = Column + lines[Line - 1].TakeWhile (a => a is ' ').Count () - 1;
+      Console.CursorLeft = cPos;
+      Console.WriteLine ("^");
+      int endPos = cPos + Text.Length / 2;
+      Console.CursorLeft = 1 + cPos - Text.Length / 2 - (endPos >= Console.WindowWidth ? Console.WindowWidth - endPos : 0);
```
Comment (tarydon) [inline Parser/Token.cs]: If Text.Length is long enough, you will end up setting Console.CursorLeft to a -ve value.

File: Parser/Token.cs [hunk 4]
```diff
@@ -37,9 +37,24 @@ public enum E {
...
-      Console.WriteLine ($"At line {Line}, column {Column} of {Source.FileName}: {Text}");
+      Console.Write ($"File: {Source.FileName}");
+      var cPos = Console.CursorLeft;
+      Console.WriteLine ($"\n\u2500\u2500\u2500\u252c{string.Join ("", Enumerable.Repeat ("\u2500", cPos - 3))}");
+      var lines = Source.Lines;
+      if (Line < lines.Length - 3) Console.WriteLine ($"{Line - 2,3}\u2502{lines[Line - 3]}");
+      if (Line < lines.Length - 2) Console.WriteLine ($"{Line - 1,3}\u2502{lines[Line - 2]}");
+      if (Line < lines.Length - 1) Console.WriteLine ($"{Line,3}\u2502{lines[Line - 1]}");
+      cPos = Column + lines[Line - 1].TakeWhile (a => a is ' ').Count () - 1;
+      Console.CursorLeft = cPos;
```
Comment (tarydon) [inline Parser/Token.cs]: Try to re-implement this without using Console.CursorLeft at all... With this approach, if the output of the compiler is redirected to a file, all the calls to Console.CursorLeft will get ignored and the output will look scrambled. Instead, you could print a ^ at the correct location by just prefixing it with a certain number of spaces.

File: Ops/Expr/GenIL.cs [hunk 5]
```diff
@@ -6,8 +6,15 @@ public class ExprILGen : Visitor<StringBuilder> {
    StringBuilder mSB = new ();
    int mID = 0;
 
-   public override StringBuilder Visit (NLiteral literal)
-      => mSB.AppendLine ($"{Label} ldc.i4 {literal.Value.Text}");
+   public override StringBuilder Visit (NLiteral literal) {
+      string load = literal.Type switch {
+         NType.Int or NType.Boolean => "ldc.i4",
+         NType.Real => "ldc.r4",
+         NType.String => "ldstr",
```
Comment (tarydon) [inline Ops/Expr/GenIL.cs:13]: Good.

File: Ops/Expr/Graph.cs [hunk 6]
```diff
@@ -24,6 +24,14 @@ public override int Visit (NIdentifier ident)
       return id; 
    }
 
+   public override int Visit (NFnCall function) {
+      List<int> args = new (); 
+      function.Params.ToList ().ForEach (a => args.Add (a.Accept (this)));
```
Comment (tarydon) [inline Ops/Expr/Graph.cs]: A neater way to do this is:
`var args = function.Params.Select (a => a.Accept(this)).ToList ()`

**Review by tarydon — CHANGES_REQUESTED** [review]

File: Ops/Expr/Graph.cs (hunk 6 above)
Comment (asfar-nazar) [reply Ops/Expr/Graph.cs]: Changed

File: Parser/Token.cs (hunk 4 above)
Comment (asfar-nazar) [reply Parser/Token.cs]: Now prefixing lines with whitespaces

File: Parser/Token.cs (hunk 2 above)
Comment (asfar-nazar) [reply Parser/Token.cs]: Added helper functions

File: Parser/Token.cs (hunk 1 above)
Comment (asfar-nazar) [reply Parser/Token.cs:56]: Corrected to print only the error text in yellow

File: Parser/Token.cs (hunk 3 above)
Comment (asfar-nazar) [reply Parser/Token.cs]: Now prefixing lines with whitespaces instead of using CursorLeft

**Pull request closed without merging**